src/game/client/w3dsnow.cpp

```diff
@@ -0,0 +1,732 @@
+            Ref_Ptr_Release(material);
+
+            if (use_point_sprites) {
+                if (!m_vertexBuffer) {
+                    Re_Acquire_Resources();
+                }
+            }
+
+            if (!use_point_sprites && !m_indexBuffer) {
+                Re_Acquire_Resources();
+            }
+
+            DX8Wrapper::Set_Texture(0, m_texture);
+
+            if (!use_point_sprites) {
+                Render_As_Quads(rinfo, x1, y1, x2, y2);
+            } else {
+                DX8Wrapper::Apply_Render_State_Changes();
+
+                DX8Wrapper::Set_DX8_Render_State(D3DRS_POINTSPRITEENABLE, TRUE);
+                DX8Wrapper::Set_DX8_Render_State(D3DRS_POINTSCALEENABLE, TRUE);
+
+                DX8Wrapper::Set_DX8_Render_State(D3DRS_POINTSIZE, F2DW(m_pointSize));
+                DX8Wrapper::Set_DX8_Render_State(D3DRS_POINTSIZE_MIN, F2DW(m_minPointSize));
+                DX8Wrapper::Set_DX8_Render_State(D3DRS_POINTSIZE_MAX, F2DW(m_maxPointSize));
+
+                DX8Wrapper::Set_DX8_Render_State(D3DRS_POINTSCALE_A, F2DW(0.0));
+                DX8Wrapper::Set_DX8_Render_State(D3DRS_POINTSCALE_B, F2DW(0.0));
+                DX8Wrapper::Set_DX8_Render_State(D3DRS_POINTSCALE_C, F2DW(1.0));
+
+                DX8Wrapper::Get_D3D_Device8()->SetStreamSource(0, m_vertexBuffer, sizeof(Vector3));
+                DX8Wrapper::Get_D3D_Device8()->SetVertexShader(DX8_FVF_XYZ);
+
+                m_vertexBufferOffset = VERTEX_COUNT;
+                m_unk5 = 45;
+                m_currentParticleCount = 0;
+
+                m_unk9 = m_amplitude + m_particleSize;
+
+                Render_Sub_Box(rinfo, x1, y1, x2, y2);
+
+                DX8Wrapper::Set_DX8_Render_State(D3DRS_POINTSPRITEENABLE, FALSE);
+                DX8Wrapper::Set_DX8_Render_State(D3DRS_POINTSCALEENABLE, FALSE);
+            }
+        }
+    }
+#endif
+}
+
+inline void W3DSnowManager::Render_Sub_Box_Internal(
+    RenderInfoClass &rinfo, MinMaxAABoxClass &minmax_aabox, int x1, int y1, int x2, int y2)
+{
+    CameraClass &camera = rinfo.m_camera;
+
+    float min_z = m_unk6 - m_boxDimensions;
+    float min_y = (float)y1 * m_boxDensity - m_unk9;
+    float min_x = (float)x1 * m_boxDensity - m_unk9;
+    minmax_aabox.m_minCorner.Set(min_x, min_y, min_z);
+
+    float max_y = (float)y2 * m_boxDensity + m_unk9;
+    float max_x = (float)x2 * m_boxDensity + m_unk9;
+    minmax_aabox.m_maxCorner.Set(max_x, max_y, m_unk6);
+
+    if (CollisionMath::Overlap_Test(camera.Get_Frustum(), AABoxClass(minmax_aabox)) != CollisionMath::OUTSIDE) {
+        Render_Sub_Box(rinfo, x1, y1, x2, y2);
+    }
+}
+
+// refactor of the code below this
+void W3DSnowManager::Render_Sub_Box(RenderInfoClass &rinfo, int x1, int y1, int x2, int y2)
+{
+#ifdef BUILD_WITH_D3D8
+    int x_delta = x2 - x1;
+    int y_delta = y2 - y1;
+
+    int x_offset = GameMath::Fast_To_Int_Ceil((x2 - x1) * 0.5);
+    int y_offset = GameMath::Fast_To_Int_Ceil((y2 - y1) * 0.5);
```

Comment:
Maybe was not written with 0.5 in original code. It is all integer, so `/ 2` is much more likely here. `0.5` is a `double`. Maybe compiler turned it into 0.5 since multiplication is cheaper than division in x86.